# Análisis Exploratorio de datos

## Proceso de ETL

### Importación de librerías

In [33]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
import json

### Funciones generales


In [2]:

def to_snake_case(text):
    text = re.sub('(.)([A-Z][a-z]+)', r'\1_/2', text)
    text = re.sub('([a-z0-9])([A-Z])', r'/1_/2', text)
    return text.lower().replace(' ', '_')


def convert_int_format(cadena):
    try:
        # Intentar convertir el string a float y luego a entero
        return str(int(float(cadena)))
    except ValueError:
        # Si no es un número, devolver el valor original
        return cadena

### Seteo de rutas

In [5]:
sheet_field_names = 'Variables'
first_row_excels= 16
sheet_dict_values = 'Valores'

'''
path_dem_adults_catalog = './Data/DataDemografica/AdolescentesAdultos/ensafrec_ad2022_entrega_w.Catlogo.xlsx'
path_dem_adults_data = './Data/DataDemografica/AdolescentesAdultos/ensafrec_ad2022_entrega_w.csv'
path_food_adults_catalog = './Data/Alimentos/AdolescentesAdultos/ensafrec_ad2022_rec_w.Catlogo.xlsx'
path_food_adults_data = './Data/Alimentos/AdolescentesAdultos/ensafrec_ad2022_rec_w.csv'


path_dem_school_catalog = './Data/DataDemografica/Escolares/ensafrec_es2022_entrega_w.Catlogo.xlsx'
path_dem_school_data = './Data/DataDemografica/Escolares/ensafrec_es2022_entrega_w.csv'
path_food_school_catalog = './Data/Alimentos/Escolares/ensafrec_es2022_rec_w.Catlogo.xlsx'
path_food_school_data = './Data/Alimentos/Escolares/ensafrec_es2022_rec_w.csv'


path_dem_preschool_catalog = './Data/DataDemografica/Preescolares/ensafrec_pr2022_entrega_w.Catlogo.xlsx'
path_dem_preschool_data = './Data/DataDemografica/Preescolares/ensafrec_pr2022_entrega_w.csv'
path_food_preschool_catalog = './Data/Alimentos/Preescolares/ensafrec_pr2022_rec_w.Catlogo.xlsx'
path_food_preschool_data = './Data/Alimentos/Preescolares/ensafrec_pr2022_rec_w.csv'

path_anthropometric_catalog = './Data/Antropometria/ensaantro2022_entrega_w.Catlogo.xlsx'
path_anthropometric_data = './Data/Antropometria/ensaantro2022_entrega_w.csv'
'''

## Data adolescentes
path_dem_adults_catalog = '.\Data\DataDemografica\AdolescentesAdultos\ensafrec_ad2022_entrega_w.Catlogo.xlsx'
path_dem_adults_data = '.\Data\DataDemografica\AdolescentesAdultos\ensafrec_ad2022_entrega_w.csv'
path_food_adults_catalog = '.\Data\Alimentos\AdolescentesAdultos\ensafrec_ad2022_rec_w.Catlogo.xlsx'
path_food_adults_data = '.\Data\Alimentos\AdolescentesAdultos\ensafrec_ad2022_rec_w.csv'

## Data escolares

path_dem_school_catalog = '.\Data\DataDemografica\Escolares\ensafrec_es2022_entrega_w.Catlogo.xlsx'
path_dem_school_data = '.\Data\DataDemografica\Escolares\ensafrec_es2022_entrega_w.csv'
path_food_school_catalog = '.\Data\Alimentos\Escolares\ensafrec_es2022_rec_w.Catlogo.xlsx'
path_food_school_data = '.\Data\Alimentos\Escolares\ensafrec_es2022_rec_w.csv'

## Data preescolares

path_dem_preschool_catalog = '.\Data\DataDemografica\Preescolares\ensafrec_pr2022_entrega_w.Catlogo.xlsx'
path_dem_preschool_data = '.\Data\DataDemografica\Preescolares\ensafrec_pr2022_entrega_w.csv'
path_food_preschool_catalog = '.\Data\Alimentos\Preescolares\ensafrec_pr2022_rec_w.Catlogo.xlsx'
path_food_preschool_data = '.\Data\Alimentos\Preescolares\ensafrec_pr2022_rec_w.csv'

## Data antrhopometrica
path_anthropometric_catalog = '.\Data\Antropometria\ensaantro2022_entrega_w.Catlogo.xlsx'
path_anthropometric_data = '.\Data\Antropometria\ensaantro2022_entrega_w.csv'




### Lectura de datos

In [6]:
#Data adolescentes
## Datos demográficos

df_dem_adults_catalog = pd.read_excel(path_dem_adults_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_dem_adults = pd.read_excel(path_dem_adults_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_dem_adults_data_raw = pd.read_csv(path_dem_adults_data, sep=';')

## Datos de alimentación

df_food_adults_catalog = pd.read_excel(path_food_adults_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_food_adults = pd.read_excel(path_food_adults_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_food_adults_data_raw = pd.read_csv(path_food_adults_data, sep=';')

#Data escolares
## Datos demográficos

df_dem_school_catalog = pd.read_excel(path_dem_school_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_dem_school = pd.read_excel(path_dem_school_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_dem_school_data_raw = pd.read_csv(path_dem_school_data, sep=';')

## Datos de alimentación

df_food_school_catalog = pd.read_excel(path_food_school_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_food_school = pd.read_excel(path_food_school_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_food_school_data_raw = pd.read_csv(path_food_school_data, sep=';')

#Data preescolares
## Datos demográficos

df_dem_preschool_catalog = pd.read_excel(path_dem_preschool_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_dem_preschool = pd.read_excel(path_dem_preschool_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_dem_preschool_data_raw = pd.read_csv(path_dem_preschool_data, sep=';')

## Datos de alimentación

df_food_preschool_catalog = pd.read_excel(path_food_preschool_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_food_preschool = pd.read_excel(path_food_preschool_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_food_preschool_data_raw = pd.read_csv(path_food_preschool_data, sep=';')

#Data antropometrica

df_anthrop_catalog = pd.read_excel(path_anthropometric_catalog, sheet_name=sheet_field_names, skiprows=first_row_excels)
df_dict_values_anthrop = pd.read_excel(path_anthropometric_catalog, sheet_name=sheet_dict_values, skiprows=first_row_excels,decimal=',')
df_anthrop_data_raw = pd.read_csv(path_anthropometric_data, sep=';')


### Procesamiento de datos

Funciones de procesamiento de datos

In [7]:
def transform_df_dictionary(df):
    # Asignar nombres a las columnas
    df.columns = ['Valor', 'ID', 'Etiqueta']
    
    
    # Aplicar la función para convertir los valores en la columna 'ID'
    df['ID'] = df['ID'].apply(convert_int_format)
    
    # Rellenar los valores faltantes en la columna 'Valor' con el valor anterior
    df.fillna(method='ffill', inplace=True)
    
    # Agrupar por 'Valor' y crear el diccionario
    df = df.groupby('Valor').apply(lambda x: dict(zip(x['ID'], x['Etiqueta']))).reset_index()
    
    # Asignar los nuevos nombres de columna
    df.columns = ['Campo', 'Diccionario']
    
    return df



def procesar_dataframe(df_data, df_catalog, df_dict):
    
    # Iterar por cada columna en df_data
    for column in df_data.columns:
        # Verificar si la columna está en el catálogo
        if column in df_catalog['Variable'].values:
            # Verificar si la columna está en el DataFrame de diccionarios
            if column in df_dict['Campo'].values:
                print(f"Procesando columna: {column}")
                # Convertir valores de la columna usando convert_int_format
                df_data[column] = df_data[column].apply(convert_int_format)
                # Mapear los valores según el diccionario correspondiente
                df_data[column] = df_data[column].map(df_dict.loc[df_dict['Campo'] == column, 'Diccionario'].values[0])
            
            # Obtener el nuevo nombre de la columna del catálogo
            new_column_name = df_catalog.loc[df_catalog['Variable'] == column, 'Etiqueta'].values[0]
            # Renombrar la columna usando to_snake_case
            df_data.rename(columns={column: to_snake_case(new_column_name)}, inplace=True)
    
    return df_data

Construcción de dataframe de diccionaros para los valores de los datos demográficos

In [8]:
df_dict_values_dem_adults = transform_df_dictionary(df_dict_values_dem_adults)
df_dict_values_dem_school = transform_df_dictionary(df_dict_values_dem_school)
df_dict_values_dem_preschool = transform_df_dictionary(df_dict_values_dem_preschool)
df_dict_values_dem_adults.head()

/var/folders/1f/sjyqyl8d1h9_ngfj8klbtf2h0000gr/T/ipykernel_58975/3060602440.py:10: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill', inplace=True)
/var/folders/1f/sjyqyl8d1h9_ngfj8klbtf2h0000gr/T/ipykernel_58975/3060602440.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df = df.groupby('Valor').apply(lambda x: dict(zip(x['ID'], x['Etiqueta']))).reset_index()
/var/folders/1f/sjyqyl8d1h9_ngfj8klbtf2h0000gr/T/ipykernel_58975/3060602440.py:10: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.f

Campo                                        Diccionario
0  completa  {'1': 'Favor de continuar (Entrevista Completa...
1   consumo         {'1': 'Igual', '2': 'Mayor', '3': 'Menor'}
2   estrato  {'1': 'Rural  ( <2500 Hab )', '2': 'Urbano ( 2...
3     nota1  {'1': '¿Acepta usted? ,,,, Sí', '2': '¿Acepta ...
4     nota2                        {'1': 'Favor de continuar'}

Ajuste de nombres de columna y valores de los datos demográficos

In [9]:
df_dem_adults_data = procesar_dataframe(df_dem_adults_data_raw, df_dem_adults_catalog, df_dict_values_dem_adults)
df_dem_school_data = procesar_dataframe(df_dem_school_data_raw, df_dem_school_catalog, df_dict_values_dem_school)
df_dem_preschool_data = procesar_dataframe(df_dem_preschool_data_raw, df_dem_preschool_catalog, df_dict_values_dem_preschool)

Procesando columna: nota1
Procesando columna: region
Procesando columna: sexo
Procesando columna: nota2
Procesando columna: consumo
Procesando columna: completa
Procesando columna: estrato
Procesando columna: nota1
Procesando columna: region
Procesando columna: sexo
Procesando columna: nota2
Procesando columna: consumo
Procesando columna: completa
Procesando columna: estrato
Procesando columna: nota1
Procesando columna: region
Procesando columna: sexo
Procesando columna: nota2
Procesando columna: consumo
Procesando columna: completa
Procesando columna: estrato


Se validan y homologan algunos nombres de columna que varían un poco

In [10]:
set(df_dem_adults_data.columns) - set(df_dem_school_data.columns)

{'frecuencia_de_consumo_en_adultos'}

In [11]:
set(df_dem_adults_data.columns) - set(df_dem_preschool_data.columns)

{'frecuencia_de_consumo_en_adultos'}

In [12]:
df_dem_adults_data.columns = ['frecuencia_de_consumo' if col.startswith('frecuencia_de_consumo') else col for col in df_dem_adults_data.columns]
df_dem_school_data.columns = ['frecuencia_de_consumo' if col.startswith('frecuencia_de_consumo') else col for col in df_dem_school_data.columns]
df_dem_preschool_data.columns = ['frecuencia_de_consumo' if col.startswith('frecuencia_de_consumo') else col for col in df_dem_preschool_data.columns]

In [13]:
df_dem_adults_data['grupo'] = 'Adolescentes y Adultos'
df_dem_school_data['grupo'] = 'Escolares'
df_dem_preschool_data['grupo'] = 'Preescolares'

Se genera un df consolidado con toda la data demográfica

In [14]:
df_dem_general = pd.concat([df_dem_adults_data, df_dem_school_data, df_dem_preschool_data], ignore_index=True)
df_dem_general.head()

folio folio_de_integrante  tiempo_hora  tiempo_minutos  \
0  2022_01001001    2022_01001001_01           20              22   
1  2022_01001028    2022_01001028_04           11              41   
2  2022_01001034    2022_01001034_03           16              33   
3  2022_01001041    2022_01001041_04           15              52   
4  2022_01001042    2022_01001042_01           15              11   

   tiempo_suma_inicio  tiempo_suma_fin hora_de__inicio_1 fecha_de__inicio_1  \
0                1212             1222          20:12:27         09/11/2022   
1                 671              701          11:11:54         09/11/2022   
2                 949              993          15:49:55         09/11/2022   
3                 937              952          15:37:59         17/11/2022   
4                 889              911          14:49:23         17/11/2022   

  hora_termino_1 fecha_termino_1  ...  duración_de_la__entrevista  \
0       20:22:36      09/11/2022  ...                          10   
1       11:41:43      09/11/2022  ...                          30   
2       16:33:37      09/11/2022  ...                          44   
3       15:52:29      17/11/2022  ...                          15   
4       15:11:30      17/11/2022  ...                          22   

   hora_de_termino fecha_de__termino  \
0         20:22:36        09/11/2022   
1         11:41:43        09/11/2022   
2         16:33:37        09/11/2022   
3         15:52:29        17/11/2022   
4         15:11:30        17/11/2022   

                   resultado_de__entrevista  \
0  Favor de continuar (Entrevista Completa)   
1  Favor de continuar (Entrevista Completa)   
2  Favor de continuar (Entrevista Completa)   
3  Favor de continuar (Entrevista Completa)   
4  Favor de continuar (Entrevista Completa)   

  especifique_otro_resultado_de_entrevista        ponderador  \
0                                            8717,1854790009   
1                                            51475,173555678   
2                                           16477,8557582843   
3                                            11801,483709865   
4                                           7032,17435561176   

      estrato_urbanidad/ruralidad  __estrato_de_seleccion  \
0  Metropolitano (100mil y + Hab)                      13   
1  Metropolitano (100mil y + Hab)                      13   
2  Metropolitano (100mil y + Hab)                      13   
3  Metropolitano (100mil y + Hab)                      13   
4  Metropolitano (100mil y + Hab)                      13   

  unidad_primaria_de_muestreo                   grupo  
0               0100100011320  Adolescentes y Adultos  
1               0100100011320  Adolescentes y Adultos  
2               0100100011320  Adolescentes y Adultos  
3               0100100011405  Adolescentes y Adultos  
4               0100100011405  Adolescentes y Adultos  

[5 rows x 57 columns]

Construcción de dataframe de diccionaros para los valores de los datos demográficos

In [15]:
df_dict_values_food_adults = transform_df_dictionary(df_dict_values_food_adults)
df_dict_values_food_school = transform_df_dictionary(df_dict_values_food_school)
df_dict_values_food_preschool = transform_df_dictionary(df_dict_values_food_preschool)
df_dict_values_food_adults.head()

/var/folders/1f/sjyqyl8d1h9_ngfj8klbtf2h0000gr/T/ipykernel_58975/3060602440.py:10: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill', inplace=True)
/var/folders/1f/sjyqyl8d1h9_ngfj8klbtf2h0000gr/T/ipykernel_58975/3060602440.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df = df.groupby('Valor').apply(lambda x: dict(zip(x['ID'], x['Etiqueta']))).reset_index()
/var/folders/1f/sjyqyl8d1h9_ngfj8klbtf2h0000gr/T/ipykernel_58975/3060602440.py:10: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.f

Campo                                        Diccionario
0  esp_alim  {'443': 'Leche en polvo entera', '445': 'Leche...
1   estrato  {'1': 'Rural  ( <2500 Hab )', '2': 'Urbano ( 2...
2       pa1  {'0': 'Nunca', '1': '1 día', '2': '2 días', '3...
3       pa2  {'1': '1 vez', '2': '2 veces', '3': '3 veces',...
4       pa3  {'1': 'POCA', '2': 'MODERADA', '3': 'MUCHA', '...

Reemplazo de valores en el DataFrame de alimentación

In [16]:
df_food_adults_data = procesar_dataframe(df_food_adults_data_raw, df_food_adults_catalog, df_dict_values_food_adults)
df_food_school_data = procesar_dataframe(df_food_school_data_raw, df_food_school_catalog, df_dict_values_food_school)
df_food_preschool_data = procesar_dataframe(df_food_preschool_data_raw, df_food_preschool_catalog, df_dict_values_food_preschool)

Procesando columna: pa1
Procesando columna: esp_alim
Procesando columna: pa2
Procesando columna: pa3
Procesando columna: estrato
Procesando columna: region
Procesando columna: pa1
Procesando columna: esp_alim
Procesando columna: pa2
Procesando columna: pa3
Procesando columna: estrato
Procesando columna: region
Procesando columna: pa1
Procesando columna: esp_alim
Procesando columna: pa2
Procesando columna: pa3
Procesando columna: estrato
Procesando columna: region


Creación de variable con el grupo al que corresponde la información

In [17]:
df_food_adults_data['grupo'] = 'Adolescentes y Adultos'
df_food_school_data['grupo'] = 'Escolares'
df_food_preschool_data['grupo'] = 'Preescolares'

Se homologan los nombres de las columnas

In [18]:
for i in range(len(df_food_adults_data.columns)):
    print(df_food_adults_data.columns[i],' - ' , df_food_school_data.columns[i], ' - ',df_food_preschool_data.columns[i])

folio  -  folio  -  folio
folio_de_integrante  -  folio_de_integrante  -  folio_de_integrante
entidad  -  entidad  -  entidad
desc_ent  -  desc_ent  -  desc_ent
municipio  -  municipio  -  municipio
desc_mun  -  desc_mun  -  desc_mun
identificador_de_alimento  -  id_alim__identificador_de_alimento  -  id_alim__identificador_de_alimento
alimento  -  alimento__alimento  -  alimento__alimento
porción_estándar  -  por_est__porción_estándar  -  por_est__porción_estándar
¿_cuántos_días_comió_(tomó)_usted?  -  pa1_¿_cuántos_días_comió_(tomó)_usted?  -  pa1_¿_cuántos_días_comió_(tomó)_usted?
especifique_otra_leche  -  esp_alim__especifique_otra_leche  -  esp_alim__especifique_otra_leche
¿_cuántas_veces_al_día_comió_(tomó)_usted?  -  pa2_¿_cuántas_veces_al_día_comió_(tomó)_usted?  -  pa2_¿_cuántas_veces_al_día_comió_(tomó)_usted?
tamaño_de_porción  -  pa3__tamaño_de_porción  -  pa3__tamaño_de_porción
número_de_porciones  -  pa4_número_de_porciones  -  pa4_número_de_porciones
ponderador  -  pond

In [90]:
df_food_school_data.columns = df_food_adults_data.columns
df_food_preschool_data.columns = df_food_adults_data.columns

Se genera un df consolidadoc con la información de alimentación

In [32]:
df_food_adults_data = df_food_adults_data.reset_index(drop=True)
df_food_school_data = df_food_school_data.reset_index(drop=True)
df_food_preschool_data = df_food_preschool_data.reset_index(drop=True)

df_food_preschool_data = df_food_preschool_data.loc[:, ~df_food_preschool_data.columns.duplicated()]

df_food_general = pd.concat([df_food_adults_data, df_food_school_data, df_food_preschool_data], ignore_index=True, axis=0)
df_food_general.head()

folio folio_de_integrante  entidad           desc_ent  municipio  \
0  2022_01001001    2022_01001001_01        1  01 AGUASCALIENTES          1   
1  2022_01001001    2022_01001001_01        1  01 AGUASCALIENTES          1   
2  2022_01001001    2022_01001001_01        1  01 AGUASCALIENTES          1   
3  2022_01001001    2022_01001001_01        1  01 AGUASCALIENTES          1   
4  2022_01001001    2022_01001001_01        1  01 AGUASCALIENTES          1   

             desc_mun  identificador_de_alimento  \
0  001 AGUASCALIENTES                        1.0   
1  001 AGUASCALIENTES                        2.0   
2  001 AGUASCALIENTES                        3.0   
3  001 AGUASCALIENTES                        4.0   
4  001 AGUASCALIENTES                        5.0   

                                            alimento porción_estándar  \
0                                      Leche Liconsa  1 vaso (240 ml)   
1                                Otra leche opción A  1 vaso (240 ml)   
2                                Otra leche opción B  1 vaso (240 ml)   
3                                Otra leche opción C  1 vaso (240 ml)   
4  Leche preparada de sabor (chocolate u otro sabor)  1 vaso (240 ml)   

  ¿_cuántos_días_comió_(tomó)_usted?  ...                   grupo  \
0                              Nunca  ...  Adolescentes y Adultos   
1                             7 días  ...  Adolescentes y Adultos   
2                              Nunca  ...  Adolescentes y Adultos   
3                              Nunca  ...  Adolescentes y Adultos   
4                              Nunca  ...  Adolescentes y Adultos   

  id_alim__identificador_de_alimento alimento__alimento  \
0                                NaN                NaN   
1                                NaN                NaN   
2                                NaN                NaN   
3                                NaN                NaN   
4                                NaN                NaN   

  por_est__porción_estándar pa1_¿_cuántos_días_comió_(tomó)_usted?  \
0                       NaN                                    NaN   
1                       NaN                                    NaN   
2                       NaN                                    NaN   
3                       NaN                                    NaN   
4                       NaN                                    NaN   

  esp_alim__especifique_otra_leche  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

   pa2_¿_cuántas_veces_al_día_comió_(tomó)_usted? pa3__tamaño_de_porción  \
0                                             NaN                    NaN   
1                                             NaN                    NaN   
2                                             NaN                    NaN   
3                                             NaN                    NaN   
4                                             NaN                    NaN   

   pa4_número_de_porciones <ninguno>  
0                      NaN       NaN  
1                      NaN       NaN  
2                      NaN       NaN  
3                      NaN       NaN  
4                      NaN       NaN  

[5 rows x 30 columns]

Se pivotea el DataFrame para tener una columna por cada alimento y una fila por cada observación

In [36]:
#with open('./Data/Inputs/food_category_data.json', 'r') as f:
with open('.\Data\Inputs\food_balance_data.json', 'r') as f:
    food_groups = json.load(f)
    df_food_general['alimento_grupo'] = df_food_general['alimento'].map(lambda x: food_groups.get(x, {}).get('groupo', 'Unknown'))
    df_food_general['alimento_subgrupo'] = df_food_general['alimento'].map(lambda x: food_groups.get(x, {}).get('subgroupo', 'Unknown'))

#with open('./Data/Inputs/food_balance_data.json', 'r') as f:
with open('.\Data\Inputs\food_balance_data.json', 'r') as f:
    food_balance = json.load(f)
    df_food_general['alimento_balance'] = df_food_general['alimento'].map(food_balance)


print( df_food_general['alimento_grupo'].nunique() )
print( df_food_general['alimento_subgrupo'].nunique() )
print( df_food_general['alimento_balance'].nunique() )


1
1
5


In [39]:
df_food_pivot = df_food_general.pivot_table(index=['grupo','folio', 'folio_de_integrante', 'entidad', 'desc_ent', 'municipio',
       'desc_mun','ponderador', 'estrato_urbanidad/ruralidad',
       '__estrato_de_seleccion', 'unidad_primaria_de_muestreo', 'completa',
       'region'], columns='alimento_grupo', values=['¿_cuántos_días_comió_(tomó)_usted?', 'especifique_otra_leche',
       '¿_cuántas_veces_al_día_comió_(tomó)_usted?', 'tamaño_de_porción',
       'número_de_porciones'], aggfunc='first')


In [40]:
df_food_pivot.shape

(983, 5)

In [94]:
df_food_pivot = df_food_pivot.reset_index()
df_food_pivot.columns = [''.join(col).strip() for col in df_food_pivot.columns.values]
df_food_pivot

grupo          folio folio_de_integrante  entidad  \
0     Adolescentes y Adultos  2022_01001001    2022_01001001_01        1   
1     Adolescentes y Adultos  2022_01001028    2022_01001028_04        1   
2     Adolescentes y Adultos  2022_01001034    2022_01001034_03        1   
3     Adolescentes y Adultos  2022_01001041    2022_01001041_04        1   
4     Adolescentes y Adultos  2022_01001042    2022_01001042_01        1   
...                      ...            ...                 ...      ...   
1579            Preescolares  2022_32052026    2022_32052026_03       32   
1580            Preescolares  2022_32053007    2022_32053007_06       32   
1581            Preescolares  2022_32053008    2022_32053008_03       32   
1582            Preescolares  2022_32053010    2022_32053010_03       32   
1583            Preescolares  2022_32053010    2022_32053010_04       32   

               desc_ent  municipio                   desc_mun  \
0     01 AGUASCALIENTES          1         001 AGUASCALIENTES   
1     01 AGUASCALIENTES          1         001 AGUASCALIENTES   
2     01 AGUASCALIENTES          1         001 AGUASCALIENTES   
3     01 AGUASCALIENTES          1         001 AGUASCALIENTES   
4     01 AGUASCALIENTES          1         001 AGUASCALIENTES   
...                 ...        ...                        ...   
1579       32 ZACATECAS         52           052 VILLA GARCÍA   
1580       32 ZACATECAS         53  053 VILLA GONZÁLEZ ORTEGA   
1581       32 ZACATECAS         53  053 VILLA GONZÁLEZ ORTEGA   
1582       32 ZACATECAS         53  053 VILLA GONZÁLEZ ORTEGA   
1583       32 ZACATECAS         53  053 VILLA GONZÁLEZ ORTEGA   

            ponderador     estrato_urbanidad/ruralidad  \
0      8717,1854790009  Metropolitano (100mil y + Hab)   
1      51475,173555678  Metropolitano (100mil y + Hab)   
2     16477,8557582843  Metropolitano (100mil y + Hab)   
3      11801,483709865  Metropolitano (100mil y + Hab)   
4     7032,17435561176  Metropolitano (100mil y + Hab)   
...                ...                             ...   
1579  6607,76451137414            Rural  ( <2500 Hab )   
1580   10209,139959835       Urbano ( 2500-99,999 Hab)   
1581  12022,6897894912       Urbano ( 2500-99,999 Hab)   
1582  12833,4574018368       Urbano ( 2500-99,999 Hab)   
1583  12022,6897894912       Urbano ( 2500-99,999 Hab)   

      __estrato_de_seleccion  ...  \
0                         13  ...   
1                         13  ...   
2                         13  ...   
3                         13  ...   
4                         13  ...   
...                      ...  ...   
1579                     321  ...   
1580                     322  ...   
1581                     322  ...   
1582                     322  ...   
1583                     322  ...   

     ¿_cuántos_días_comió_(tomó)_usted?Yogur en vaso: Bajo en grasa o light natural o con fruta (vitalínea, alpura light, lala light, etcétera)  \
0                                                   NaN                                                                                           
1                                                   NaN                                                                                           
2                                                   NaN                                                                                           
3                                                   NaN                                                                                           
4                                                   NaN                                                                                           
...                                                 ...                                                                                           
1579                                                NaN                                                                                           
1580       

Se crea un df para tener la información de las porciones de los alimentos

In [95]:
df_portions = df_food_general[['identificador_de_alimento', 'alimento', 'porción_estándar']].drop_duplicates()
df_portions.head()

identificador_de_alimento  \
0                          1   
1                          2   
2                          3   
3                          4   
4                          5   

                                            alimento porción_estándar  
0                                      Leche Liconsa  1 vaso (240 ml)  
1                                Otra leche opción A  1 vaso (240 ml)  
2                                Otra leche opción B  1 vaso (240 ml)  
3                                Otra leche opción C  1 vaso (240 ml)  
4  Leche preparada de sabor (chocolate u otro sabor)  1 vaso (240 ml)

Las sigueintes columnas son númericas, por lo que no aplica el reemplazo de valores, sin embargo se debe tener en cuenta que cuando se pone 222 o 999 es porque no se sabe el valor real, por lo que se debe tener en cuenta para el análisis.

In [96]:
antr_numerical_columns = ['an01_1', 'an01_2', 'an04_1', 'an04_2', 'an07', 'an12_1', 'an12_2', 'an15_1', 'an15_2','an17_1','an19','an21_1','an21_2','an23','an25','an27_01s',
                          'an27_01d', 'an27_02s', 'an27_02d', 'an27_03s', 'an27_03d'  ]

Construcción de dataframe de diccionaros para los valores de los datos demográficos

In [97]:
df_dict_values_anthrop = transform_df_dictionary(df_dict_values_anthrop)
df_dict_values_anthrop.head()

Campo                                        Diccionario
0  an01_1                              {'222': 'No se pesó'}
1  an01_2                              {'222': 'No se pesó'}
2    an02  {'1': 'Ligera', '2': 'Gruesa', '3': 'Sin ropa'...
3    an03  {'1': 'Sin problema', '2': 'Problema físico', ...
4  an04_1                             {'222': 'No se midió'}

In [98]:
df_dict_values_anthrop = df_dict_values_anthrop[~df_dict_values_anthrop['Campo'].isin(antr_numerical_columns)]

Ajuste de nombres de columna y valores de los datos antropometricos

In [99]:
df_antroph_data = procesar_dataframe(df_anthrop_data_raw, df_anthrop_catalog, df_dict_values_anthrop)

Procesando columna: nota1
Procesando columna: region
Procesando columna: h0302
Procesando columna: nota01
Procesando columna: an02
Procesando columna: an03
Procesando columna: nota02
Procesando columna: an05
Procesando columna: an06
Procesando columna: nota03
Procesando columna: an09
Procesando columna: nota04
Procesando columna: an10
Procesando columna: an11
Procesando columna: an13
Procesando columna: an14
Procesando columna: an16
Procesando columna: an18
Procesando columna: an20
Procesando columna: an22
Procesando columna: nota05
Procesando columna: an24
Procesando columna: an26
Procesando columna: nota06
Procesando columna: an29
Procesando columna: an30
Procesando columna: completa
Procesando columna: estrato


In [100]:
df_antroph_data.head()

folio folio_de_integrante  tiempo_hora  tiempo_minutos  \
0  2022_01001001    2022_01001001_01           20              31   
1  2022_01001001    2022_01001001_02           20              28   
2  2022_01001003    2022_01001003_02           20              36   
3  2022_01001003    2022_01001003_03           19              29   
4  2022_01001003    2022_01001003_06           20              37   

   tiempo_suma_inicio  tiempo_suma_fin hora_de__inicio_1 fecha_de__inicio_1  \
0                1228             1231          20:28:39         09/11/2022   
1                1226             1228          20:26:30         09/11/2022   
2                1231             1236          20:31:35         09/11/2022   
3                1080             1169          18:00:46         10/11/2022   
4                1236             1237          20:36:38         09/11/2022   

  hora_termino_1 fecha_termino_1  ...  \
0       20:31:33      09/11/2022  ...   
1       20:28:37      09/11/2022  ...   
2       20:36:37      09/11/2022  ...   
3       19:29:56      10/11/2022  ...   
4       20:37:59      09/11/2022  ...   

                ¿_algún_comentario_de_la_entrevista?  \
0                                                      
1                                                      
2                                                      
3  ta correcta de la 3° es de 89/43 , se coloca 8...   
4                                                      

   duración_de_la__entrevista hora_de_termino fecha_de__termino  \
0                           3        20:31:33        09/11/2022   
1                           2        20:28:37        09/11/2022   
2                           5        20:36:37        09/11/2022   
3                          89        19:29:56        10/11/2022   
4                           1        20:37:59        09/11/2022   

  resultado_de__entrevista especifique_otro_resultado_de_entrevista  \
0      ENTREVISTA COMPLETA                                            
1      ENTREVISTA COMPLETA                                            
2      ENTREVISTA COMPLETA                                            
3      ENTREVISTA COMPLETA                                            
4      ENTREVISTA COMPLETA                                            

         ponderador     estrato_urbanidad/ruralidad __estrato_de_seleccion  \
0  999,761285509631  Metropolitano (100mil y + Hab)                     13   
1  1546,78591945017  Metropolitano (100mil y + Hab)                     13   
2  2926,13359214152  Metropolitano (100mil y + Hab)                     13   
3  1819,85902017304  Metropolitano (100mil y + Hab)                     13   
4   4640,3577583505  Metropolitano (100mil y + Hab)                     13   

  unidad_primaria_de_muestreo  
0               0100100011320  
1               0100100011320  
2               0100100011320  
3               0100100011320  
4               0100100011320  

[5 rows x 99 columns]

### Dataframes finales

In [70]:
df_food_pivot.head()

grupo          folio folio_de_integrante  entidad  \
0  Adolescentes y Adultos  2022_01001001    2022_01001001_01        1   
1  Adolescentes y Adultos  2022_01001028    2022_01001028_04        1   
2  Adolescentes y Adultos  2022_01001034    2022_01001034_03        1   
3  Adolescentes y Adultos  2022_01001041    2022_01001041_04        1   
4  Adolescentes y Adultos  2022_01001042    2022_01001042_01        1   

            desc_ent  municipio            desc_mun        ponderador  \
0  01 AGUASCALIENTES          1  001 AGUASCALIENTES   8717,1854790009   
1  01 AGUASCALIENTES          1  001 AGUASCALIENTES   51475,173555678   
2  01 AGUASCALIENTES          1  001 AGUASCALIENTES  16477,8557582843   
3  01 AGUASCALIENTES          1  001 AGUASCALIENTES   11801,483709865   
4  01 AGUASCALIENTES          1  001 AGUASCALIENTES  7032,17435561176   

      estrato_urbanidad/ruralidad  __estrato_de_seleccion  ...  \
0  Metropolitano (100mil y + Hab)                      13  ...   
1  Metropolitano (100mil y + Hab)                      13  ...   
2  Metropolitano (100mil y + Hab)                      13  ...   
3  Metropolitano (100mil y + Hab)                      13  ...   
4  Metropolitano (100mil y + Hab)                      13  ...   

  ¿_cuántos_días_comió_(tomó)_usted?Yogur en vaso: Bajo en grasa o light natural o con fruta (vitalínea, alpura light, lala light, etcétera)  \
0                                                NaN                                                                                           
1                                                NaN                                                                                           
2                                                NaN                                                                                           
3                                                NaN                                                                                           
4                                                NaN                                                                                           

   ¿_cuántos_días_comió_(tomó)_usted?Yogur en vaso: Entero con frutas  \
0                                                NaN                    
1                                                NaN                    
2                                                NaN                    
3                                                NaN                    
4                                                NaN                    

  ¿_cuántos_días_comió_(tomó)_usted?Yogur en vaso: Entero natural  \
0                                                NaN                
1                                                NaN                
2                                                NaN                
3                                                NaN                
4                                                NaN                

  ¿_cuántos_días_comió_(tomó)_usted?Yogur para beber: Bajo en grasa o light natural o con fruta (vitalínea, activia 0%, Siluette, etcétera)  \
0                                                NaN                                                                                          
1                                                NaN                                                                                          
2                                                NaN                                                                                          
3                                                NaN                                                                                          
4                                                NaN                                                                                          

  ¿_cuántos_días_comió_(tomó)_usted?Yogur para beber: Entero con frutas  \
0                                                NaN                      
1                

In [71]:
df_antroph_data.head()

folio folio_de_integrante  tiempo_hora  tiempo_minutos  \
0  2022_01001001    2022_01001001_01           20              31   
1  2022_01001001    2022_01001001_02           20              28   
2  2022_01001003    2022_01001003_02           20              36   
3  2022_01001003    2022_01001003_03           19              29   
4  2022_01001003    2022_01001003_06           20              37   

   tiempo_suma_inicio  tiempo_suma_fin hora_de__inicio_1 fecha_de__inicio_1  \
0                1228             1231          20:28:39         09/11/2022   
1                1226             1228          20:26:30         09/11/2022   
2                1231             1236          20:31:35         09/11/2022   
3                1080             1169          18:00:46         10/11/2022   
4                1236             1237          20:36:38         09/11/2022   

  hora_termino_1 fecha_termino_1  ...  \
0       20:31:33      09/11/2022  ...   
1       20:28:37      09/11/2022  ...   
2       20:36:37      09/11/2022  ...   
3       19:29:56      10/11/2022  ...   
4       20:37:59      09/11/2022  ...   

                ¿_algún_comentario_de_la_entrevista?  \
0                                                      
1                                                      
2                                                      
3  ta correcta de la 3° es de 89/43 , se coloca 8...   
4                                                      

   duración_de_la__entrevista hora_de_termino fecha_de__termino  \
0                           3        20:31:33        09/11/2022   
1                           2        20:28:37        09/11/2022   
2                           5        20:36:37        09/11/2022   
3                          89        19:29:56        10/11/2022   
4                           1        20:37:59        09/11/2022   

  resultado_de__entrevista especifique_otro_resultado_de_entrevista  \
0      ENTREVISTA COMPLETA                                            
1      ENTREVISTA COMPLETA                                            
2      ENTREVISTA COMPLETA                                            
3      ENTREVISTA COMPLETA                                            
4      ENTREVISTA COMPLETA                                            

         ponderador     estrato_urbanidad/ruralidad __estrato_de_seleccion  \
0  999,761285509631  Metropolitano (100mil y + Hab)                     13   
1  1546,78591945017  Metropolitano (100mil y + Hab)                     13   
2  2926,13359214152  Metropolitano (100mil y + Hab)                     13   
3  1819,85902017304  Metropolitano (100mil y + Hab)                     13   
4   4640,3577583505  Metropolitano (100mil y + Hab)                     13   

  unidad_primaria_de_muestreo  
0               0100100011320  
1               0100100011320  
2               0100100011320  
3               0100100011320  
4               0100100011320  

[5 rows x 99 columns]

In [72]:
df_dem_general.head()

folio folio_de_integrante  tiempo_hora  tiempo_minutos  \
0  2022_01001001    2022_01001001_01           20              22   
1  2022_01001028    2022_01001028_04           11              41   
2  2022_01001034    2022_01001034_03           16              33   
3  2022_01001041    2022_01001041_04           15              52   
4  2022_01001042    2022_01001042_01           15              11   

   tiempo_suma_inicio  tiempo_suma_fin hora_de__inicio_1 fecha_de__inicio_1  \
0                1212             1222          20:12:27         09/11/2022   
1                 671              701          11:11:54         09/11/2022   
2                 949              993          15:49:55         09/11/2022   
3                 937              952          15:37:59         17/11/2022   
4                 889              911          14:49:23         17/11/2022   

  hora_termino_1 fecha_termino_1  ...  ¿_algún_comentario_de_la_entrevista?  \
0       20:22:36      09/11/2022  ...                                         
1       11:41:43      09/11/2022  ...                               NINGUNO   
2       16:33:37      09/11/2022  ...                               NINGUNO   
3       15:52:29      17/11/2022  ...                                         
4       15:11:30      17/11/2022  ...                                         

   duración_de_la__entrevista hora_de_termino fecha_de__termino  \
0                          10        20:22:36        09/11/2022   
1                          30        11:41:43        09/11/2022   
2                          44        16:33:37        09/11/2022   
3                          15        15:52:29        17/11/2022   
4                          22        15:11:30        17/11/2022   

                   resultado_de__entrevista  \
0  Favor de continuar (Entrevista Completa)   
1  Favor de continuar (Entrevista Completa)   
2  Favor de continuar (Entrevista Completa)   
3  Favor de continuar (Entrevista Completa)   
4  Favor de continuar (Entrevista Completa)   

  especifique_otro_resultado_de_entrevista        ponderador  \
0                                            8717,1854790009   
1                                            51475,173555678   
2                                           16477,8557582843   
3                                            11801,483709865   
4                                           7032,17435561176   

      estrato_urbanidad/ruralidad __estrato_de_seleccion  \
0  Metropolitano (100mil y + Hab)                     13   
1  Metropolitano (100mil y + Hab)                     13   
2  Metropolitano (100mil y + Hab)                     13   
3  Metropolitano (100mil y + Hab)                     13   
4  Metropolitano (100mil y + Hab)                     13   

  unidad_primaria_de_muestreo  
0               0100100011320  
1               0100100011320  
2               0100100011320  
3               0100100011405  
4               0100100011405  

[5 rows x 56 columns]

In [73]:
df_portions.head()

identificador_de_alimento  \
0                          1   
1                          2   
2                          3   
3                          4   
4                          5   

                                            alimento porción_estándar  
0                                      Leche Liconsa  1 vaso (240 ml)  
1                                Otra leche opción A  1 vaso (240 ml)  
2                                Otra leche opción B  1 vaso (240 ml)  
3                                Otra leche opción C  1 vaso (240 ml)  
4  Leche preparada de sabor (chocolate u otro sabor)  1 vaso (240 ml)

FRECUENCIA DE CONSUMO EN ADULTOS = 
ID_ALIM Identificador de alimento = 

Pendientes:
- Revisar aleatoriamente si el procesamiento quedó bien


Ideas:
- Crear grupos de alimentos para que no sean tantos

Notas corrida:
- En algunos archivos del catálogo los últimos dos campos no tienen etiqueta, se deben poner de manera manual en el Excel.